### DISCUSSION TITLE
Using pyvista to create a volume of polygonal surfaces

### DISCUSSION THREAD

**Discussion by liberostelios** [pyvista/pyvista#1429] [Q&A]
We are using `pyvista` to compute stats for building objects from [CityJSON](https://www.cityjson.org) (you can see code [here](https://github.com/liberostelios/star-wars)). In CityJSON, objects can have surfaces that are bound by complex polygons (ie more than 4 vertices); but as far as we are concerned `pyvista` seems to be providing odd results when passing faces other than triangles and quads.

An example is the following building from a dataset (first as shown in Blender, then as plotted by `pyvista`):
<img width="358" alt="Screenshot 2021-06-23 at 15 52 37" src="https://user-images.githubusercontent.com/1896327/123108832-0bbc9780-d43b-11eb-9291-1ee9812f8258.png">
<img width="479" alt="Screenshot 2021-06-23 at 15 52 52" src="https://user-images.githubusercontent.com/1896327/123108869-1545ff80-d43b-11eb-9b09-a0aaf43cff11.png">

Is there a way to represent these complex surfaces in `pyvista` or does it only work with triangles and quads? If not, then maybe there is an easy way to triangulate these surfaces?

**Comment by liberostelios** [comment]
A small update on the matter: it turns out that if you call the triangulate filter it just fixes the surfaces:

<img width="669" alt="Screenshot 2021-06-25 at 12 50 13" src="https://user-images.githubusercontent.com/1896327/123414203-e56e3780-d5b3-11eb-99c4-9fcb9bd61660.png">

I am not sure if this counts as an issue that could be fixed so that a triangulation is not necessary.

**Reply by akaszynski** [reply]
Triangulation shouldn't be necessary as `pyvista` and VTK support faces with more than 4 edges.  It's possible the data isn't being read into `pyvista` properly and triangulating it cleans it up.  Have you tried using the [clean](https://docs.pyvista.org/core/filters.html#pyvista.PolyDataFilters.clean) filter?

**Reply by liberostelios** [reply]
Thanks a lot for the response.

I double checked the code and the data seem to be read properly. I actually found some cases where even the triangulation doesn't work.

Here is a building non-triangulated (and cleaned):
<img width="762" alt="Screenshot 2021-06-27 at 12 09 05" src="https://user-images.githubusercontent.com/1896327/123540672-7d009100-d740-11eb-85f6-a73750bb5904.png">

Same building triangulated (and cleaned):
<img width="725" alt="Screenshot 2021-06-27 at 12 10 26" src="https://user-images.githubusercontent.com/1896327/123540707-ade0c600-d740-11eb-9d0e-1387b1722966.png">

I extracted only one side that is problematic there. If I plot the indices as a line I get this:
```python
pts = side.points
idx = [6, 1, 0, 4, 3, 2, 5]

sface = pv.PolyData(pts)
sface.lines = idx
sface.plot()
```
<img width="433" alt="Screenshot 2021-06-27 at 12 11 03" src="https://user-images.githubusercontent.com/1896327/123540726-c224c300-d740-11eb-9f01-ed365c8613f1.png">

But the exact same indices as a face provide this:
```python
pts = side.points
idx = [6, 1, 0, 4, 3, 2, 5]

sface = pv.PolyData(pts, idx)
sface.clean().plot(show_edges=True)
```
<img width="433" alt="Screenshot 2021-06-27 at 12 11 53" src="https://user-images.githubusercontent.com/1896327/123540749-e1235500-d740-11eb-86c6-95257cbb72d3.png">

**Reply by banesullivan** [reply]
Can you share a minimal example? For example, take one of the problematic meshes and save it: `mesh.save('mesh.vtk')` for us to try out?
